File: src/04.stack/05.227.ts
export function calculate(s: string): number {
    const n = s.length;
    const stack: number[] = [];
    let sign = "+";
    let num = 0;

    for (let i = 0; i < n; i++) {
        if (/[0-9]/.test(s[i])) {
            num = 10 * num + (s[i].charCodeAt(0) - "0".charCodeAt(0));
        }

        if ((!/[0-9]/.test(s[i]) && s[i] !== " ") || i === n - 1) {
            switch (sign) {
                case "+":
                    stack.push(num);
                    break;
                case "-":
                    stack.push(-num);
                    break;
                case "*":
                    let pre = stack.pop();
                    stack.push(pre * num);
                    break;
                case "/":
                    let pre1 = stack.pop();
                    stack.push(Math.trunc(pre1 / num));
                    break;
            }
            num = 0;
            sign = s[i];
        }
    }

    let res = 0;
    for (let i = 0; i < stack.length; i++) {
        res += stack[i];
    }

    return res;
}

File: src/04.stack/05.227.go
package stack

import "unicode"

func Calculate(s string) int {
	n := len(s)
	stack := []int{}
	num := 0
	sign := '+'

	for i, v := range s {
		if unicode.IsDigit(v) {
			num = num*10 + int(v-'0')
		}

		if !unicode.IsDigit(v) && v != ' ' || i == n-1 {
			switch sign {
			case '+':
				stack = append(stack, num)
			case '-':
				stack = append(stack, -num)
			case '*':
				stack[len(stack)-1] *= num
			case '/':
				stack[len(stack)-1] /= num
			default:
				break
			}
			sign = v
			num = 0
		}
	}

	res := 0

	for _, v := range stack {
		res += v
	}

	return res
}
